src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -91,34 +108,128 @@ def self.compile_ast(code, compile_options, boltlib_path)
             # to compile a catalog in that context from the supplied AST.
-            pal.with_catalog_compiler do |compiler|
-              # We have to parse the AST inside the compiler block, because it
-              # initializes the necessary type loaders for us.
-              ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+            pal.with_catalog_compiler(**variables) do |compiler|
+              # TODO: PUP-10476 Explore setting these as default in PAL. They are the defaults in Puppet
+              Puppet[:strict] = :warning
+              Puppet[:strict_variables] = false
 
+              ast = build_program(code)
               compiler.evaluate(ast)
+              compiler.evaluate_ast_node
               compiler.compile_additions
               compiler.catalog_data_hash
             end
           end
         end
       end
-      private_class_method :compile_ast
+      private_class_method :compile_for_environment_plan
+
+      def self.compile_for_environment(code, compile_options, boltlib_path)
+        # When we do not need to load bolt we assume there are no Bolt types to resolve in
+        # AST compilation.
+        Puppet[:node_name_value] = compile_options['certname']
+
+        # Use the existing environment with the requested name
+        Puppet::Pal.in_environment(compile_options['environment'],
+                                    envpath: Puppet[:environmentpath],
+                                    facts: compile_options['facts']['values'],
+                                    variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            # We have to parse the AST inside the compiler block, because it
+            # initializes the necessary type loaders for us.
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+
+            compiler.evaluate(ast)
+            compiler.compile_additions
+            compiler.catalog_data_hash
+          end
+        end
+      end
+      private_class_method :compile_for_environment
+
+      def self.with_puppet_settings(overrides = {})
+        # Create a copy of the existing settings object
+        original_settings = Puppet.settings
+        original_basemodulepath = Puppet[:basemodulepath]
+        original_vendormoduledir = Puppet[:vendormoduledir]
+
+        Dir.mktmpdir('bolt_compile') do |dir|
+          puppet_config = {}
+          Puppet::Settings::REQUIRED_APP_SETTINGS.each do |setting|
+            puppet_config[setting] = dir
+          end
+          Puppet::Server::PuppetConfig.initialize_puppet(puppet_config: puppet_config)
```

Comment:
I know you are working on a new way of safely setting puppet settings. I think in bolt we use that clear_all_settings_for_test method. Anyway, for this context this blows up:
```
2021-07-22T18:04:27.336Z ERROR [qtp1366641407-62] [p.r.core] Internal Server Error: org.jruby.exceptions.RuntimeError: (DevError) Attempting to initialize global default settings more than once!
        at RUBY.initialize_global_settings(/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/settings.rb:301)
        at RUBY.do_initialize_settings_for_run_mode(/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:184)
        at RUBY.initialize_settings(/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:159)
        at RUBY.initialize_puppet_settings(uri:classloader:/puppetserver-lib/puppet/server/puppet_config.rb:35)
        at RUBY.initialize_puppet(uri:classloader:/puppetserver-lib/puppet/server/puppet_config.rb:69)
        at RUBY.with_puppet_settings(uri:classloader:/puppetserver-lib/puppet/server/ast_compiler.rb:214)
        at RUBY.mktmpdir(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/tmpdir.rb:93)
        at RUBY.with_puppet_settings(uri:classloader:/puppetserver-lib/puppet/server/ast_compiler.rb:209)
        at RUBY.compile_for_project_plan(uri:classloader:/puppetserver-lib/puppet/server/ast_compiler.rb:95)
        at RUBY.compile_ast(uri:classloader:/puppetserver-lib/puppet/server/ast_compiler.rb:38)
        at RUBY.compile(uri:classloader:/puppetserver-lib/puppet/server/ast_compiler.rb:21)
        at RUBY.compileAST(uri:classloader:/puppetserver-lib/puppet/server/master.rb:104)
```